File: 100/105 - Construct Binary Tree from Preorder and Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder.length != inorder.length)
            return null;
        
        return helper(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1);
    }
    
    private TreeNode helper(int[] preorder, int preFrom, int preTo, int[] inorder, int inFrom, int inTo) {
        TreeNode root;
        int index = -1, i;
        
        if (preTo < preFrom || inTo < inFrom)
            return null;
        root = new TreeNode(preorder[preFrom]);
        for (i = inFrom; i <= inTo; ++i) {
            if (inorder[i] == root.val) {
                index = i;
                break;
            }
        }
        if (index < 0)
            return null;
        
        root.left = helper(preorder, preFrom + 1, preFrom + (index - inFrom), inorder, inFrom, index - 1);
        root.right = helper(preorder, preFrom + (index - inFrom + 1), preTo, inorder, index + 1, inTo);
        
        return root;
    }
}

File: csharp/105.construct-binary-tree-from-preorder-and-inorder-traversal.cs
/*
 * @lc app=leetcode id=105 lang=csharp
 *
 * [105] Construct Binary Tree from Preorder and Inorder Traversal
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {

    private TreeNode BuildTreeRecursive(int[] preorder, int[] inorder, int pStart, int pEnd, int iStart, int iEnd) {
        if (pStart > pEnd) {
            return null;
        }
        int val = preorder[pStart];
        TreeNode ret = new TreeNode(val);
        int leftEnd = iStart - 1;
        while (leftEnd < iEnd) {
            if (inorder[leftEnd + 1] == val) {
                break;
            }
            ++leftEnd;
        }
        int leftLen = leftEnd - iStart + 1;
        ret.left = BuildTreeRecursive(preorder, inorder, pStart + 1, pStart + leftLen, iStart, leftEnd);
        ret.right = BuildTreeRecursive(preorder, inorder, pStart + leftLen + 1, pEnd, leftEnd + 2, iEnd);
        return ret;
    }

    public TreeNode BuildTree(int[] preorder, int[] inorder) {
        int len = preorder.Length;
        return BuildTreeRecursive(preorder, inorder, 0, len - 1, 0, len - 1);
    }
}

